samples/matter/common/src/board_interface.cpp

```diff
@@ -0,0 +1,358 @@
+			EventManager::PostEvent(timerEvent);
+		},
+		reinterpret_cast<intptr_t>(timer));
+}
+
+void BoardInterface::FunctionTimerEventHandler(const void *context)
+{
+	if (!context) {
+		return;
+	}
+	SystemEvent event(context);
+
+	/* If we reached here, the button was held past kFactoryResetTriggerTimeout, initiate factory reset */
+	if (sInstance.mFunction == SystemEventType::SoftwareUpdate) {
+		LOG_INF("Factory reset has been triggered. Release button within %ums to cancel.",
+			FactoryResetConsts::kFactoryResetTriggerTimeout);
+
+		/* Start timer for kFactoryResetCancelWindowTimeout to allow user to cancel, if required. */
+		sInstance.StartTimer(FactoryResetConsts::kFactoryResetCancelWindowTimeout);
+		sInstance.mFunction = SystemEventType::FactoryReset;
+
+		/* Turn off all LEDs before starting blink to make sure blink is coordinated. */
+		sInstance.ResetAllLeds();
+
+		sInstance.mStatusLED.Blink(LedConsts::kBlinkRate_ms);
+		sInstance.mApplicationLED.Blink(LedConsts::kBlinkRate_ms);
+#if NUMBER_OF_LEDS == 4
+		sInstance.mUserLED1.Blink(LedConsts::kBlinkRate_ms);
+		sInstance.mUserLED2.Blink(LedConsts::kBlinkRate_ms);
+#endif
+	} else if (sInstance.mFunction == SystemEventType::FactoryReset) {
+		/* Actually trigger Factory Reset */
+		sInstance.mFunction = SystemEventType::None;
+		chip::Server::GetInstance().ScheduleFactoryReset();
+
+	} else if (sInstance.mFunction == SystemEventType::AdvertisingStart) {
+		/* The button was held past kAdvertisingTriggerTimeout, start BLE advertisement
+		   if we have 2 buttons UI*/
+#if NUMBER_OF_BUTTONS == 2
+		StartBLEAdvertisement();
+		sInstance.mFunction = SystemEventType::None;
+#endif
+	}
+}
+
+void BoardInterface::ButtonEventHandler(uint32_t buttonState, uint32_t hasChanged)
+{
+	SystemEvent buttonEvent;
+	bool isAppButtonEvent = false;
+	ButtonActions action;
+	DeviceButtons source;
+
+	buttonEvent.mType = static_cast<uint8_t>(SystemEventType::Button);
+
+	if (BLUETOOTH_ADV_BUTTON_MASK & hasChanged) {
+		buttonEvent.ButtonEvent.PinNo = BLUETOOTH_ADV_BUTTON;
+		buttonEvent.ButtonEvent.Action = static_cast<uint8_t>((BLUETOOTH_ADV_BUTTON_MASK & buttonState) ?
+									      SystemEventType::ButtonPushed :
+									      SystemEventType::ButtonReleased);
+		buttonEvent.mHandler = StartBLEAdvertisementHandler;
+		EventManager::PostEvent(buttonEvent);
+	}
+
+	if (FUNCTION_BUTTON_MASK & hasChanged) {
+		buttonEvent.ButtonEvent.PinNo = FUNCTION_BUTTON;
+		buttonEvent.ButtonEvent.Action =
+			static_cast<uint8_t>((FUNCTION_BUTTON_MASK & buttonState) ? SystemEventType::ButtonPushed :
+										    SystemEventType::ButtonReleased);
+		buttonEvent.mHandler = FunctionHandler;
+		EventManager::PostEvent(buttonEvent);
+	}
+
+#if NUMBER_OF_BUTTONS == 4
+	if (APPLICATION_BUTTON_MASK & hasChanged) {
+		source = DeviceButtons::kAppButton;
+		action = (APPLICATION_BUTTON_MASK & buttonState) ? ButtonActions::kButtonPressed :
+								   ButtonActions::kButtonReleased;
+		isAppButtonEvent = true;
```

Comment:
Note that DK buttons and LEDs module supports registering multiple handlers via https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/libraries/others/dk_buttons_and_leds.html#c.dk_button_handler_add, so an application could just register an extra handler to take care of app-specific buttons.